### ISSUE TITLE
Cannot save "required" dynamic-zone or component

### ISSUE THREAD

**Issue by Torsten85** [strapi/strapi#17071]
### Required System information

- Node.js version: ^18
- NPM version: ^8
- Strapi version: 4.11.2
- Database: sqlite
- Operating system: macOS
- Is your project Javascript or Typescript: Javascript

### Describe the bug

Setting a dynamic-zone or component (repeatable or single) to "required" makes it impossible to publish a document.
Strapi does not detect a filled dynamic-zone / component as valid.

### Steps to reproduce the behavior

1. Fresh strapi install
2. Create a content type with a dynamic zone
3. Make that dynamic zone required
4. Create a new document
5. Add stuff to the dynamic zone
6. see error when saving / publishing

### Expected behavior

Strapi should show an error when the dynamic-zone / component is empty.
Strapi should show no error when the dynamic-zone /component has components.

**Comment by drazik** [comment]
I have the same problem on a totally fresh Strapi instance. In the following screenshot, the repeatable component is obviously not empty, but when I try to publish the content, I get  an error:

![image](https://github.com/strapi/strapi/assets/1606068/a3062b28-59d9-4069-9968-6c3ed73dc325)

**Comment by lamuertepeluda** [comment]
Same problem here: 😢 

* Node.js version: ^18
* Yarn version: 1.22.19
* Strapi version: 4.11.2
* Database: Postgres 14
* Operating system: macOS
* Is your project Javascript or Typescript: Typescript

**Comment by lamuertepeluda** [comment]
Trying to debug, I see there is a validation error. 

The POST to the publish endpoint http://localhost:1337/content-manager/collection-types/{uid}/{id}/actions/publish from the Admin UI when pressing the Publish Button
returns a 400 bad request with "<required dynamic-zone field name> must be defined" as a message.

I only get this error when _publishing_. Creating or updating my Collection Type (with Save) does not trigger any problem.

I have inspected the package at tag v4.11.2 with the node.js debugger as well, and when this occurs it happens that the dynamic zone fields are not populated before calling the validator at 
packages/core/content-manager/server/services/entity-manager.js:152

i.e `await strapi.entityValidator.validateEntityCreation`

While when calling the validation within create() or update() methods, those are populated

**Comment by lamuertepeluda** [comment]
As an additional note, I am using a super admin user

**Comment by lamuertepeluda** [comment]
Additionally, the fact that I don't get the problem on create and update is probably due to a customization/override I had to do, on src/extensions/content-manager/strapi-server.ts

The problem, as far as I can see, is that something as changed in the collection types controller, in node_modules/@strapi/plugin-content-manager/server/controllers/collection-types.js

The lines causing trouble are 195-200,

```javascript
    const permissionQuery = await permissionChecker.sanitizedQuery.publish(ctx.query);
    const populate = await getService('populate-builder')(model)
      .populateFromQuery(permissionQuery)
      .build();

    const entity = await entityManager.findOne(id, model, { populate });
```

after line 200, entity does not contain the dynamic zone fields. Previously this was managed differently as

```javascript
const entity = await entityManager.findOneWithCreatorRoles(id, model);
```

In fact, I had to adapt my override as well since `findOneWithCreatorRoles` has been removed (though I cannot keep track of this in the changelogs).

Hope this helps the team identifying this problem faster.
I think in the meanwhile I am going to override the publish method as well 🤦🏼

**Comment by lamuertepeluda** [comment]
I guess the problem is in how the `populate` variable above gets built.
I sadly have no time to investigate it deeper, but replacing the populate with the following code:

```js
    const entity = await entityManager.findOne(id, model, {
      populate: getDeepPopulate(model), // <- Patch here
    });
```

patches the problem. I think the approach the strapi team was attempting was a little more subtle, but probably because of some bug is not working.

The trick is finding the `getDeepPopulate` function in the package utils 🙄 ...

@Torsten85 and @drazik I share my override here (["Plugin Extension"](https://docs.strapi.io/dev-docs/plugins-extension) in the Strapi world), which seems to work, hoping it helps until this gets fixed. Code is in TS, but I guess JS should look quite close. 

Create a file in `src/extensions/content-manager/strapi-server.ts` (or .js and replace `import` with `require()` and remove type annotations)

```typescript
import strapiUtils from "@strapi/utils";
import { getService } from "@strapi/plugin-content-manager/server/utils";
import { getDeepPopulate } from "@strapi/plugin-content-manager/server/services/utils/populate";

// This will override the content-manager core plugin (only methods that need override)
export default (plugin) => {
  // check node_modules/@strapi/plugin-content-manager/server/controllers/collection-types.js
  const controller = plugin.controllers["collection-types"];

  /**
   * Override publish to mitigate https://github.com/strapi/strapi/issues/17071
   * @param ctx
   * @returns
   */
  controller.publish = async (ctx) => {
    const { userAbility, user } = ctx.state;
    const { id, model } = ctx.params;

    const entityManager = getService("entity-manager");
    const permissionChecker = getService("permission-checker").create({
      userAbility,
      model,
    });

    if (permissionChecker.cannot.publish()) {
      return ctx.forbidden();
    }

    // PART OF CODE CAUSING TROUBLES
    // const permissionQuery = await permissionChecker.sanitizedQuery.publish(
    //   ctx.query
    // );
    // const populate = await getService("populate-builder")(model)
    //   .populateFromQuery(permissionQuery)
    //   .build();

    const entity = await entityManager.findOne(id, model, {
      populate: getDeepPopulate(model), // <- Patch here
    });

    if (!entity) {
      return ctx.notFound();
    }

    if (permissionChecker.cannot.publish(entity)) {
      return ctx.forbidden();
    }

    const result = await entityManager.publish(
      entity,
      model,
      strapiUtils.setCreatorFields({ user, isEdition: true })({})
    );

    ctx.body = await permissionChecker.sanitizeOutput(result);
  };

  return plugin;
};


```

**Comment by webholics** [comment]
I have the same problem with required relations in the latest Strapi v4.11.2.

**Comment by Marc-Roig** [comment]
Hello, I am reviewing this big and will provide a fix asap. It seems to only happen on publish.

**Comment by Marc-Roig** [comment]
The fix has been merged and will be available in tomorrow's 4.11.3 release 🚀  Thank you all for the detailed report of the issue!!
